### PR TITLE
ggml: share native MTP compute buffers and synchronize catch-up

### PR DESCRIPTION
**Pull request by amangupta-tether**
## Overview

Native MTP currently reserves a second compute workspace even when the target and draft run sequentially. Port https://github.com/ggml-org/llama.cpp/pull/27489 onto `temp-10549`, preserving this fork's scheduler configuration and the original author attribution. Schedulers retain independent allocation plans while sharing reference-counted backing buffers; larger reservations detach.

Also synchronize the MTP catch-up decode before `process()` returns. `llama_decode(ctx_dft, ...)` is asynchronous, so without this barrier the next target batch can overwrite shared workspace while draft kernels are still running. The barrier also runs on the catch-up error path.

Compute sharing now requires the explicit `ctx_other_share_compute` opt-in (default false); the serialized common MTP path opts in. Sharing remains limited to the same model object, one sequence, one compute device (CPU/ACCEL helpers excluded), and compatible scheduler layouts. A separately loaded MTP head retains independent buffers.

## Additional information

Validated on NVIDIA GB10 with CUDA 13.0.88 against base `57ef02cb4`, using `Qwen3.8-27B-UD-Q8_K_L.gguf`. The separate-head comparison uses the same model's extracted MTP block and required global tensors; all 18 tensor payloads were verified byte-for-byte by SHA-256.

Peak live GGML CUDA backing-buffer allocations (MiB):

| Context / KV | MTP head | Base | Patch | Saved |
|---|---|---:|---:|---:|
| 8192 / f16 | Embedded | 26700.60 | 26570.59 | 130.02 |
| 8192 / f16 | Separate | 28323.65 | 28323.65 | 0.00 |
| 196608 / q4_0 | Embedded | 31624.60 | 30596.59 | 1028.02 |
| 196608 / q4_0 | Separate | 33247.65 | 33247.65 | 0.00 |

Measured with an external allocation/free tracer that counts each physical GGML CUDA buffer once. Includes model, KV, and compute buffers; excludes pinned host memory, backend-internal pools, and driver allocations. These figures are not total GPU memory usage.

Validation:

- CUDA Release build and `test-alloc` pass, including the ported shared-buffer lifecycle test.
- One slot, all layers offloaded, batch/microbatch 512, Flash Attention on, maximum draft length 2, eight CPU threads.
- Each base/patched configuration generated 128 tokens from a 16-token prompt and 128 from a 1536-token prompt spanning three prefill batches, with temperature 0 and seed 1234. All 16 completions succeeded; output hashes and accepted/generated draft counts matched in every base/patch pair.
- All runs exited cleanly and released every traced CUDA buffer.
- For the 1536-token prompt with the embedded head, generation was 13.39 -> 13.03 tok/s at 8K/f16 and 13.95 -> 13.61 tok/s at 196608/q4_0. These are single-run observations, not a statistically controlled performance result.

The large context setting tests reservation size; the prompt itself contains 1536 tokens.

### Vulkan and Metal validation

Sharing is backend-independent: there is no backend-name allowlist. Eligibility requires the explicit context opt-in, matching model identity, one sequence, one compute device (CPU/ACCEL helpers excluded), matching scheduler devices and buffer types, and a single scheduler copy. Target/MTP execution must be serialized.

Tested on `qvac-dev-strix-1` (Radeon 8060S / RADV Mesa 25.2.8) with cached `Qwen3.6-27B-MTP-Q4_K_M.gguf`, and `qvac-dev-mac-arm64` (M3 Ultra) with cached `Qwen_Qwen3.5-35B-A3B-Q4_0.gguf`. These backend tests use different models from the Qwen3.8-27B CUDA tests above.

Peak GPU compute backing allocations, MiB (excludes model/KV/host buffers and backend-private or driver allocations):

| Backend | Context / KV | Independent | Shared | Saved |
|---|---|---:|---:|---:|
| metal | 8192/f16 | 167.03 | 84.57 | 82.46 |
| metal | 196608/q4_0 | 1271.14 | 636.63 | 634.52 |
| vulkan | 8192/f16 | 272.15 | 142.13 | 130.02 |
| vulkan | 196608/q4_0 | 590.15 | 326.13 | 264.02 |

Both settings used one sequence, batch/microbatch 512, Flash Attention, and draft maximum 2. Baseline was the CUDA-only version of this PR, including its synchronization fix. All 16 additional completions (short and 1536-token multi-batch prompts, 128 generated tokens each) succeeded, with identical output hashes and accepted/generated draft counts in every within-backend comparison. All tracked compute buffers were freed. Allocator tests passed on both backends, and the extended gate also builds and passes `test-alloc` with CUDA.

Memory was measured with identical temporary allocation instrumentation in both remote builds; that instrumentation is not included in this PR. The large context setting tests reservation capacity, not a full-length prompt.

### Review fixes and current GB10 validation (3908bbeba)

Sharing is now explicit at the context API. A weak scheduler-state reference removes the target-context lifetime dependency. Scheduler generations reattach the draft after target sampler changes; compatible target rebuilds retain their workspace. CPU and ACCEL helpers do not count as additional compute devices.

Added synthetic target/MTP execution regressions covering asynchronous completion (including an error injected after backend submission), multiple prefill batches, graph reuse, sampler attach/remove, ACCEL helpers, and freeing the target before draft reservation. CUDA CTest passes all three selected tests. The CPU-backed sharing variant and allocator tests pass AddressSanitizer with leak detection. Removing the catch-up barrier makes the regression fail. The CUDA regression is registered in the `main` CTest suite used by GPU CI.

Retested on this NVIDIA GB10 with `Qwen3.8-27B-UD-Q8_K_L.gguf`. Each of eight configurations generated 128 tokens from a short prompt, 128 from a 1536-token prompt spanning three prefill batches, then repeated the long request. One sequence, all layers offloaded, batch/microbatch 512, Flash Attention on, MTP maximum 2, eight CPU threads, temperature 0 and seed 1234. All 24 completions succeeded, paired output hashes and draft counts matched, and all traced buffers were released.

Physical GGML CUDA backing allocations, MiB (same scope as the original CUDA table):

| Context / KV | Backend sampling | Independent after requests | Shared after requests | Saved | Independent peak | Shared peak |
|---|---|---:|---:|---:|---:|---:|
| 8192 / f16 | off | 26700.60 | 26570.59 | 130.02 | 26700.60 | 26570.59 |
| 196608 / q4_0 | off | 31624.60 | 30596.59 | 1028.02 | 31624.60 | 30596.59 |
| 8192 / f16 | on | 26700.60 | 26570.59 | 130.02 | 26700.60 | 26700.60 |
| 196608 / q4_0 | on | 31624.60 | 30596.59 | 1028.02 | 31624.60 | 31624.60 |

After-request memory stayed stable across all three requests, including backend sampling. With backend sampling, the first sampler grows the workspace slightly; safe detachment temporarily keeps both workspaces alive before the draft reattaches. Therefore the sampling-on savings apply after requests, while sampling-off savings also reduce whole-run peak allocations. The independent comparison uses the same binary and external tracing, with an API interceptor clearing the sharing opt-in. No tracing instrumentation is included in the PR. These figures are not total GPU memory usage.

## Requirements

- [x] I have read and agree with the [contributing guidelines](https://github.com/tetherto/qvac-fabric-llm.cpp/blob/temp-10549/CONTRIBUTING.md).
- AI usage disclosure: YES - Codex assisted with porting the upstream change, adding synchronization, extracting the comparison head, running measurements, and preparing this PR.

### CODE REVIEW

File: src/llama-context.cpp [hunk 1]
```diff
@@ -659,6 +665,25 @@ void llama_context::sched_reserve() {
...
+            ggml_backend_dev_t device = ggml_backend_get_device(backend);
+            if (device == nullptr || ggml_backend_dev_type(device) == GGML_BACKEND_DEVICE_TYPE_CPU) {
+                continue;
+            }
+
+            n_devices++;
+        }
+
+        if (n_devices == 1 &&
+            ggml_backend_sched_share_compute_buffers(sched.get(), ctx_compute->get_sched())) {
```
Comment (DmitryMalishev) [inline src/llama-context.cpp]: Major: enforce serialization before automatically sharing compute buffers. Matching model identity and scheduler layouts does not ensure that target and draft execution cannot overlap. Sequential calls to llama_decode(draft) and llama_decode(target) can still leave asynchronous work running on separate backend streams, allowing one context to overwrite the other's workspace. The barrier in common_speculative_impl_draft_mtp::process() protects that caller, but direct users of the public context API bypass it. Make sharing explicitly opt-in for callers that guarantee serialization, or enforce synchronization between sharing contexts.

File: tests/test-alloc.cpp [hunk 2]
```diff
@@ -583,6 +663,102 @@ static void test_reallocation() {
     }
 }
 
+static void test_shared_buffers() {
```
Comment (DmitryMalishev) [inline tests/test-alloc.cpp:666]: Major: add an automated regression test for asynchronous target/MTP execution. These dummy-backend tests cover allocation, detachment, ownership, and reset-callback rejection, but never execute asynchronous target/draft work. Removing the new catch-up synchronization barrier would leave them passing. Add an alternating target/MTP execution test covering multiple prefill batches, graph reuse, and error recovery, and run it in GPU CI.

File: src/llama-context.cpp [hunk 3]
```diff
@@ -162,6 +162,12 @@ llama_context::llama_context(
         }
     }
 
+    if (params.ctx_type == LLAMA_CONTEXT_TYPE_MTP && params.ctx_other != nullptr &&
```
Comment (gianni-cor) [inline src/llama-context.cpp]: ctx_compute is an unowned raw pointer to params.ctx_other. After the target context is freed, a later draft sched_reserve() still calls ctx_compute->get_sched() and hits a heap-use-after-free. Reproduced with ASAN on spark-d9ca at this call in sched_reserve(). Either keep the target alive for the MTP context lifetime, or clear ctx_compute when the target is destroyed.

File: src/llama-context.cpp [hunk 4]
```diff
@@ -659,6 +665,25 @@ void llama_context::sched_reserve() {
     };
     create_sched(cparams.pipeline_parallel);
 
+    if (ctx_compute != nullptr) {
+        int n_devices = 0;
```
Comment (gianni-cor) [inline src/llama-context.cpp:681]: n_devices counts every non-CPU backend, including GGML_BACKEND_DEVICE_TYPE_ACCEL. On qvac-dev-mac-arm64 with Accelerate, backend_ptrs.size() = 3 (Metal, CPU, BLAS) so sharing is skipped (compute buffer sharing unavailable). Without Accelerate, backend_ptrs.size() = 2 and sharing works. Ignore ACCEL helpers here; they are not a second compute device.

File: src/llama-context.cpp (hunk 1 above)
Comment (gianni-cor) [inline src/llama-context.cpp]: Sharing is attached only when the draft scheduler is created. llama_set_sampler() recreates the target scheduler, so the draft keeps the old workspace and the target allocates a new one. With -bs / backend sampling on spark-d9ca, each request reallocates a separate 130 MiB CUDA compute buffer and sharing is lost. Re-share after target scheduler recreation, or do not recreate the target scheduler while an MTP context still points at it.

Comment (amangupta-tether) [reply src/llama-context.cpp]: Fixed in [3908bbeba](https://github.com/tetherto/qvac-fabric-llm.cpp/commit/3908bbeba845e5a5429a75a891dac6ee0bb3fb46). Sharing now requires the explicit `llama_context_params::ctx_other_share_compute` opt-in, which defaults to false. Its API contract requires serialization of all sharing contexts, including completion of asynchronous work before switching contexts. The common MTP initializer opts in and retains the catch-up synchronization barrier. Direct context API users keep independent workspaces unless they explicitly accept that contract. The regression checks that the default is false and that non-opted-in target/draft graphs do not alias.

File: tests/test-alloc.cpp (hunk 2 above)
Comment (amangupta-tether) [reply tests/test-alloc.cpp:666]: Added in [3908bbeba](https://github.com/tetherto/qvac-fabric-llm.cpp/commit/3908bbeba845e5a5429a75a891dac6ee0bb3fb46): `test-mtp-shared` executes a synthetic Qwen35 target/MTP pair on CUDA, compares shared versus independent logits, and checks multiple prefill batches, actual graph reuse, sampler attach/remove, recovery after an error injected after backend submission, and target destruction followed by draft reservation. A backend probe asserts that `process()` leaves no submitted draft work pending, including on the error path. Temporarily removing the catch-up barrier makes that assertion fail. The test is registered with CTest's `main` label for CUDA builds, so the existing self-hosted GPU CI runs it. Local CUDA CTest passed; the CPU-backed variant and allocator tests also passed AddressSanitizer with leak detection.

File: src/llama-context.cpp (hunk 3 above)
Comment (amangupta-tether) [reply src/llama-context.cpp]: Fixed in [3908bbeba](https://github.com/tetherto/qvac-fabric-llm.cpp/commit/3908bbeba845e5a5429a75a891dac6ee0bb3fb46). The draft no longer stores a raw target-context pointer. It holds a weak reference to scheduler state owned by the target; target destruction expires that state, so a later draft reservation cannot dereference the freed context and falls back to independent allocation. Physical buffer ownership remains reference-counted. The new regression explicitly frees the target before rebuilding the draft scheduler and passes both on CUDA and under AddressSanitizer using a CPU-backed compute device.

File: src/llama-context.cpp (hunk 4 above)
Comment (amangupta-tether) [reply src/llama-context.cpp:681]: Fixed in [3908bbeba](https://github.com/tetherto/qvac-fabric-llm.cpp/commit/3908bbeba845e5a5429a75a891dac6ee0bb3fb46). The device count now excludes both CPU and ACCEL helpers, while scheduler device/buffer-type compatibility and reset-callback rejection remain in place. The new sharing regression registers an ACCEL helper alongside the compute backend and CPU fallback, and asserts actual target/draft buffer aliasing. That passes locally with CUDA and with the CPU-backed compute fixture. This validates the gate without claiming a new Metal/Accelerate hardware run.

File: src/llama-context.cpp (hunk 1 above)
Comment (amangupta-tether) [reply src/llama-context.cpp]: Fixed in [3908bbeba](https://github.com/tetherto/qvac-fabric-llm.cpp/commit/3908bbeba845e5a5429a75a891dac6ee0bb3fb46). Target scheduler rebuilds now try to retain the existing workspace through the shared allocator. A scheduler generation counter makes the draft rebuild its allocation plan and reattach to the current target workspace before its next decode. Sampler-driven rebuilds remain supported. The regression verifies sharing after sampler attach/remove.

Retested on this GB10 with `Qwen3.8-27B-UD-Q8_K_L.gguf`, `-bs`, and three requests per run (short, 1536-token prefill, repeated 1536-token prefill; 128 generated tokens each). Savings after every request are 130.02 MiB at 8192/f16 and 1028.02 MiB at 196608/q4_0. Output hashes and accepted/generated draft counts match independent-buffer runs; all traced buffers are freed.

Peak qualification: the first backend sampler grows the workspace slightly, so safe detachment temporarily retains both workspaces. With `-bs`, whole-run peak backing allocations remain approximately equal to the independent run; the savings above are after requests. Without `-bs`, both peak and post-request savings remain 130.02/1028.02 MiB.

**Review by DmitryMalishev — APPROVED** [review]
